File: ios/Sources/MeasureSDK/Swift/Tracing/InvalidSpan.swift
//
//  InvalidSpan.swift
//  Measure
//
//  Created by Adwin Ross on 08/04/25.
//

import Foundation

/// An invalid span implementation that does nothing.
final class InvalidSpan: Span {
    let traceId: String = "invalid-trace-id"
    let spanId: String = "invalid-span-id"
    let isSampled: Bool = false
    let parentId: String? = nil

    func setStatus(_ status: SpanStatus) -> Span { return self }
    func setParent(_ parentSpan: Span) -> Span { return self }
    func setCheckpoint(_ name: String) -> Span { return self }
    func setName(_ name: String) -> Span { return self }
    func setAttribute(_ key: String, value: String) -> Span { return self }
    func setAttribute(_ key: String, value: Int) -> Span { return self }
    func setAttribute(_ key: String, value: Double) -> Span { return self }
    func setAttribute(_ key: String, value: Bool) -> Span { return self }
    func setAttributes(_ attributes: [String: Any]) -> Span { return self }
    func removeAttribute(_ key: String) -> Span { return self }
    func end() -> Span { return self }
    func end(timestamp: Number) -> Span { return self }
    func hasEnded() -> Bool { return false }
    func getDuration() -> Int64 { return 0 }
}

File: react-native/src/tracing/invalidSpan.ts
import type { ValidAttributeValue } from "../utils/attributeValueValidator";
import type { Span } from "./span";
import type { SpanStatus } from "./spanStatus";

/**
 * An invalid span implementation that does nothing (No-Op Span).
 * Used when tracing is disabled or a valid span could not be created.
 */
export class InvalidSpan implements Span {
    traceId: string = "invalid-trace-id";
    spanId: string = "invalid-span-id";
    isSampled: boolean = false;
    parentId: string | undefined = undefined;
    
    setStatus(_status: SpanStatus): Span { return this; }
    setParent(_parentSpan: Span): Span { return this; }
    setCheckpoint(_name: string): Span { return this; }
    setName(_name: string): Span { return this; }
    setAttribute(_key: string, _value: string | number | boolean): Span { return this; }
    setAttributes(_attributes: { [key: string]: ValidAttributeValue }): Span { return this; }
    removeAttribute(_key: string): Span { return this; }
    end(_timestampMs?: number): Span { return this; }
    hasEnded(): boolean { return false; }
    getDuration(): number { return 0; }
}